lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,204 @@
+import { enableLogging, logDebugMessage } from "../logger";
+
+import type { SuperTokensNextjsConfig, SuperTokensRequestToken } from "./types";
+
+const ACCESS_TOKEN_COOKIE_NAME = "sAccessToken";
+const ACCESS_TOKEN_HEADER_NAME = "st-access-token";
+const FRONT_TOKEN_COOKIE_NAME = "sFrontToken";
+const FRONT_TOKEN_HEADER_NAME = "front-token";
+const REFRESH_TOKEN_COOKIE_NAME = "sRefreshToken";
+const REFRESH_TOKEN_HEADER_NAME = "st-refresh-token";
+const ANTI_CSRF_TOKEN_COOKIE_NAME = "sAntiCsrf";
+const ANTI_CSRF_TOKEN_HEADER_NAME = "anti-csrf";
+
+const REDIRECT_ATTEMPT_MAX_COUNT = 5;
+const REDIRECT_PATH_PARAM_NAME = "stRedirectTo";
+const FORCE_LOGOUT_PATH_PARAM_NAME = "forceLogout";
+const REDIRECT_ATTEMPT_COUNT_COOKIE_NAME = "sSsrSessionRefreshAttempt";
+
+let AppInfo: SuperTokensNextjsConfig["appInfo"];
+
+export async function refreshSession(config: SuperTokensNextjsConfig, request: Request): Promise<Response> {
+    AppInfo = config.appInfo;
+    if (config.enableDebugLogs) {
+        enableLogging();
+    }
+
+    // Cancel the refresh cycle if an unforseen state is encountered
+    const redirectAttemptNumber = getRedirectAttemptNumber(request);
+    if (redirectAttemptNumber > REDIRECT_ATTEMPT_MAX_COUNT) {
+        return redirectToAuthPage(request);
+    }
+
+    const refreshToken =
+        getCookie(request, REFRESH_TOKEN_COOKIE_NAME) || request.headers.get(REFRESH_TOKEN_HEADER_NAME);
+    if (!refreshToken) {
+        logDebugMessage("Refresh token not found");
+        return redirectToAuthPage(request);
+    }
+
+    const requestUrl = new URL(request.url);
+    const redirectTo = requestUrl.searchParams.get(REDIRECT_PATH_PARAM_NAME) || "/";
+    try {
+        const tokens = await fetchNewTokens(refreshToken);
+        const hasRequiredCookies = tokens.accessToken.cookie && tokens.refreshToken.cookie && tokens.frontToken.cookie;
+        const hasRequiredHeaders = tokens.accessToken.header && tokens.refreshToken.header && tokens.frontToken.header;
+        if (!hasRequiredCookies && !hasRequiredHeaders) {
+            logDebugMessage("Missing tokens from refresh response");
+            return redirectToAuthPage(request);
+        }
+        const redirectUrl = new URL(redirectTo, request.url);
```

Comment:
Do we check if redirectTo is relative?

Reply:
I don't think we need to. Since we are the ones setting it in the SSR function